nums/core/array/application.py

```diff
@@ -332,28 +331,37 @@ def diag(self, X: BlockArray) -> BlockArray:
                 syskwargs = {"grid_entry": grid_entry, "grid_shape": grid.grid_shape}
                 if np.all(np.diff(grid_entry) == 0):
                     # This is a diagonal block.
-                    rarr.blocks[grid_entry].oid = self.cm.diag(X.blocks[grid_entry[0]].oid,
+                    rarr.blocks[grid_entry].oid = self.cm.diag(X.blocks[grid_entry[0]].oid, 0,
                                                                syskwargs=syskwargs)
                 else:
                     rarr.blocks[grid_entry].oid = self.cm.new_block("zeros",
                                                                     grid_entry,
                                                                     grid_meta,
                                                                     syskwargs=syskwargs)
         elif len(X.shape) == 2:
-            assert X.shape[0] == X.shape[1], "X must be a square array."
-            assert X.block_shape[0] == X.block_shape[1], "block_shape must be square."
-            shape = X.shape[0],
-            block_shape = X.block_shape[0],
-            grid = ArrayGrid(shape, block_shape, X.dtype.__name__)
-            rarr = BlockArray(grid, self.cm)
-            for grid_entry in X.grid.get_entry_iterator():
-                out_grid_entry = grid_entry[:1]
-                out_grid_shape = grid.grid_shape[:1]
-                syskwargs = {"grid_entry": out_grid_entry, "grid_shape": out_grid_shape}
-                if np.all(np.diff(grid_entry) == 0):
-                    # This is a diagonal block.
-                    rarr.blocks[out_grid_entry].oid = self.cm.diag(X.blocks[grid_entry].oid,
-                                                                   syskwargs=syskwargs)
+            out_shape = min(X.shape),
+            out_block_shape = min(X.block_shape),
+            # Obtain the block indices which contain the diagonal of the matrix.
+            diag_meta = array_utils.find_output_blocks(X.blocks, out_shape[0])
+
+            all_result_blocks = []
+            out_grid_shape, count = (len(diag_meta),), 0
```

Comment:
Please separate these into two lines. In general, define variables on separate lines. It took me multiple scans to identify where `count` was being defined.

Reply:
Done. Sorry about that.